src/core/Database.cpp

```diff
@@ -116,14 +116,15 @@ bool Database::open(QSharedPointer<const CompositeKey> key, QString* error, bool
  * @param error error message in case of failure
  * @return true on success
  */
-bool Database::open(const QString& filePath, QSharedPointer<const CompositeKey> key, QString* error, bool readOnly)
+KeePass2Reader::Status
+Database::open(const QString& filePath, QSharedPointer<const CompositeKey> key, QString* error, bool readOnly)
```

Comment:
I really don't like the Database class returning the status type of a completely different class. That is the point of the error string to capture the actual error and to either succeed in opening (true) or fail (false). Having more than those states complicates handling of hard/soft failures.

Reply:
My thought was that it would make the API more transparent and explicit, because you know that this is a case you may want to handle. But you don't have to, the enum can be cast implicitly to bool if you are interested only in hard failures. This status isn't needed, I could read it from the Database object, but it would make an edge case more obvious.

I am unsure about whether we want this or not. It may make things more complicated, but it would definitely make them more explicit. Also returning an error code in combination with an error description is generally a useful thing to do (most C/C++ libraries do exactly that), because you cannot handle different errors or warnings appropriately if there's only a description in natural language.

If you are concerned only about the "other class" thing I could duplicate the enum to this class.

Reply:
> This status isn't needed, I could read it from the Database object
I was thinking that as well, keep the reader simple and out of these concerns. 

Its nice to detect and report on an edge case situation. We just need to make sure that 1) the user even cares, 2) we might just want to silently fix the problem or issue a warning message, and/or 3) just hard fail when a version mismatch exists (I don't recommend).

Reply:
I don't really care about the user of the application here. I'm thinking more of users of our API, who are forced to think about why there's an enum returned.

Reply:
I moved the `hasMinorVersionMismatch()` method into the `Database` class and reverted it back to a bool. That's still not as explicit, but more explicit than completely ignoring the fact that there might be a minor version mismatch.

Reply:
Part of me wants to make a DatabaseFileHandler class eventually and move all the open/save into it. Technically the KeePass2Reader should return a QSharedPointer<Database> that is subsequently used instead of writing into a provided Database instance. Almost all classes would interact with the DatabaseFileHandler class.

Reply:
TBH, I would rather have Database class with a load() method that manages the allocated resources RAII-style and then have it call a bunch of free functions for doing the actual parsing and loading stuff. Just because it's Qt, doesn't mean we have to do everything OOP (thank goodness).